Fix nsn starting with national prefix

Alex-Usmanov opened the pull request:
When removing a national prefix in national number parser, sometimes the removal does not occur (and further down the line, when transform rule is applied, weird behavior is encountered).

In order to remedy this (and allow for type-as-you-go formatting), a measure is necessary to remove the previously applied transform rule (leadingDigits from metadata do not reflect transform rule all the time - example isocode=cc). So, I added a mechanism for removal of this transform rule (which is usually the national prefix).

cedvdb commented on lib/src/parsers/_national_number_parser.dart:
```diff
@@ -11,11 +11,29 @@ abstract class NationalNumberParser {
-    if (nationalPrefix == null) return nationalNumber;
-    if (nationalNumber.startsWith(nationalPrefix)) {
-      return nationalNumber.substring(nationalPrefix.length);
+    final transformPrefix = patterns.nationalPrefixTransformRule
```
could you explain this line a bit ?

cedvdb reviewed (commented):
Could you add a test that would have been failing before the fix to show what was the problem more clearly, and to be sure it does not break in the future ?

Alex-Usmanov commented:
Hi!
Sorry for late response.

Turns out the issue was slightly different that I anticipated (moreover, some tests were failing in my version of solution).
So, the issue is: when formatting a national number, a transformRule might be applied when it is unnecessary.

For instance we shall take this number: 4416666
When parsing it with `destinationCountry=IsoCode.BM`, the transform rule is applied and 4416666 turns into 4414416666.

To fix this, I added a check to the transform function to see if the number is already transformed and do not make the transformation again.

Added a test case:
`phone_number_test.dart:190`

cedvdb commented on lib/src/parsers/_national_number_parser.dart:
```diff
@@ -35,6 +40,14 @@ abstract class NationalNumberParser {
     final transformRule = patterns.nationalPrefixTransformRule;
 
+    final transformPrefix = patterns.nationalPrefixTransformRule
+        ?.replaceAllMapped(RegExp(r'(\$\d)'), (_) => '');
```
could you add an example of what this would do for your personal case as a comment above this line ? Thanks

Alex-Usmanov replied:
Added more comments ✅  (in slightly different place however, I could manage to fit it in here - the meaning did not match with the code)

cedvdb commented:
Sorry for the late reply from myself too.

At first glance I've the following worries:

 1. It seems hacky to parse the rule pattern like that
 2. There are  `nationalPrefixFormattingRule` of the sort ` r"$NP$FG"` which I'm not sure how that'd interact with the given code.

Maybe an alternative solution would be to only apply the transform rule when the phone number is valid ?
